### PR TITLE
Various refactorings to share more code with future generators

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Refactor a lot of code out of LibraryImportGenerator.csproj and into Microsoft.Interop.SourceGeneration.csproj so we can reuse more of it with the upcoming COM source generator.

This PR moves many of the building blocks for the way the interop team has designed our source generators into a more shareable design and makes most of LibraryImportGenerator into glue between the building blocks and special LibraryImportGenerator-only features.

### CODE REVIEW

**Comment by jkoritzinsky** [comment]
This PR also removes an old warning that only fires in "unsupported" scenarios that will never come up for consumers (custom CoreLib assembly) that caused additional complexity.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/GeneratedStatements.cs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        public ImmutableArray<StatementSyntax> Setup { get; init; }
+        public ImmutableArray<StatementSyntax> Marshal { get; init; }
+        public ImmutableArray<FixedStatementSyntax> Pin { get; init; }
+        public StatementSyntax InvokeStatement { get; init; }
+        public ImmutableArray<StatementSyntax> Unmarshal { get; init; }
+        public ImmutableArray<StatementSyntax> KeepAlive { get; init; }
+        public ImmutableArray<StatementSyntax> GuaranteedUnmarshal { get; init; }
+        public ImmutableArray<StatementSyntax> Cleanup { get; init; }
+
+        public static GeneratedStatements GenerateStatements(BoundGenerators marshallers, StubCodeContext context, ExpressionSyntax expressionToInvoke)
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/GeneratedStatements.cs:27]: Maybe just `Generate`? `GeneratedStatements.GenerateStatements` seems a bit redundant.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/SignatureContext.cs [hunk 2]
```diff
@@ -82,53 +73,21 @@ public IEnumerable<ParameterSyntax> StubParameters
...
-
-            (ImmutableArray<TypePositionInfo> typeInfos, IMarshallingGeneratorFactory generatorFactory) = GenerateTypeInformation(method, libraryImportData, diagnostics, env);
+            ImmutableArray<TypePositionInfo> typeInfos = GenerateTypeInformation(method, libraryImportData, diagnostics, env);
 
             ImmutableArray<AttributeListSyntax>.Builder additionalAttrs = ImmutableArray.CreateBuilder<AttributeListSyntax>();
 
             if (env.TargetFramework != TargetFramework.Unknown)
             {
+                string GeneratorName = generatorInfoAssembly.GetName().Name;
+                string GeneratorVersion = generatorInfoAssembly.GetName().Version.ToString();
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/SignatureContext.cs:90]: `GeneratorName` -> `generatorName`
`GeneratorVersion` -> `generatorVersion`

**Review by elinor-fung — COMMENTED** [review]
I really just looked at how `PInvokeStubCodeGenerator` now composes the generated code - which seems reasonable to me. I'm assuming all most of the added code is just copied/moved with slight tweaks to fit nicely in their new place - is there a specific part we should look at closely?

**Comment by jkoritzinsky** [comment]
Yeah all of the code here was just moved around. No big logic changes in particular anywhere.

**Review by elinor-fung — APPROVED** [review]

**Pull request merged**